DiHydrogen, Hydrogen, and Aluminum CachedCMakePackage

bvanessen opened the pull request:
Switching the DiHydrogen, Hydrogen, and Aluminum packages to use CachedCMakePackage.

Removed deprecated options, variants, and versions that should no longer be supported.

Updated the LBANN package to interface with the new packages.

Added cmake_cache_filepath and fixed the CMAKE_PREFIX_PATH to return a string not a path, since it is really a list of paths, but not actually a path.

bvanessen committed:
Switching the DiHydrogen, Hydrogen, and Aluminum packages to use
CachedCMakePackage.

Removed deprecated options, variants, and versions that should no
longer be supported.

Updated the LBANN package to interface with the new packages.

Added cmake_cache_filepath and fixed the CMAKE_PREFIX_PATH to return a
string not a path, since it is really a list of paths, but not
actually a path.

bvanessen committed:
Updated Aluminum to properly depend on RCCL. Removed duplicate helper function.

haampie commented:
Can you also look into why `lbann` started picking up `dihydrogen@develop`? This happens in Gitlab CI.

It's going to be fixed/disallowed by #39611, but I think using `dihydrogen@develop` was unintentional

bvanessen committed:
Moving common code patterns to the base CachedCMakePackage class.

bvanessen committed:
Update the AMD GPU target propogation in Aluminum.

bvanessen committed:
Add hooks to find the llvm-amdgpu binarie in the cached cmake packages.

bvanessen committed:
Put back in the limit for cmake on earlier versions.

bvanessen committed:
Cleaned up the LBANN recipe and dropped deprecated versions.

Removed unnecessary variants for packages that are now standard
dependencies.

Added version 1.5.3 for Hydrogen and set it as a new minimum version
for LBANN.

Set more variants to sticky to avoid the concretizer toggling
significant variants.

bvanessen committed:
Added flag to ease debugging.

bvanessen commented:
@spackbot fix style

bvanessen committed:
[@spackbot] updating style on behalf of bvanessen

bvanessen committed:
Fixed Flake8

bvanessen committed:
Fixed more Flake8

benson31 reviewed (approved):
Looks fine to me. (I only reviewed the package recipes as that's my primary interest here.)

bvanessen committed:
Updated versions of LBANN and DiHydrogen.  Removed references to deprecated Aluminum versions.

bvanessen committed:
Re-added the dependency on py-protobuf+cc

bvanessen committed:
Push Aluminum library to master when LBANN is at develop.

bvanessen commented:
> Can you also look into why `lbann` started picking up `dihydrogen@develop`? This happens in Gitlab CI.
> 
> It's going to be fixed/disallowed by #39611, but I think using `dihydrogen@develop` was unintentional

Probably because of the set of version constraints for different packages.  We have released new versions, so hopefully this is fixed.

bvanessen commented:
@haampie This PR is ready to go.

bvanessen committed:
Added a patch to fix some build issues seen on certain compilers.

haampie reviewed (approved):
I just glanced over it, assuming you know what you're doing...

bvanessen commented:
@haampie Thank you.

adrienbernede commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -257,6 +262,15 @@ def initconfig_hardware_entries(self):
+            # others point to /<path>/rocm-<ver>/llvm
+            if os.path.basename(os.path.normpath(llvm_prefix)) != "llvm":
+                llvm_bin = os.path.join(llvm_prefix, "llvm/bin/")
+            entries.append(
+                cmake_cache_filepath("CMAKE_HIP_COMPILER", os.path.join(llvm_bin, "clang++"))
+            )
```
@bvanessen do you remember the exact explanation for this ?
I have been in the middle of a revamp in #39613 and now in #41375 instead, and I am trying to find the right way of supporting all the cases.
My understanding is that we need this for recent rocm builds, when the spack spec is asking for +rocm without a proper amdgpu compiler.
However, this is co-existing with the "old fashion way", which is to set HIP_CXX_COMPILER to hipcc (see line 263) and I wonder in it should be the case.
